src/spikeinterface/preprocessing/detect_bad_channels.py

```diff
@@ -4,8 +4,122 @@
+    The standard deviation/mad multiplier threshold
+psd_hf_threshold : float, default: 0.02
+    For coherence+psd - an absolute threshold (uV^2/Hz) used as a cutoff for noise channels.
+    Channels with average power at >80% Nyquist larger than this threshold
+    will be labeled as noise
+dead_channel_threshold : float, default: -0.5
+    For coherence+psd - threshold for channel coherence below which channels are labeled as dead
+noisy_channel_threshold : float, default: 1
+    Threshold for channel coherence above which channels are labeled as noisy (together with psd condition)
+outside_channel_threshold : float, default: -0.75
+    For coherence+psd - threshold for channel coherence above which channels at the edge of the recording are marked as outside
+    of the brain
+outside_channels_location : "top" | "bottom" | "both", default: "top"
+    For coherence+psd - location of the outside channels. If "top", only the channels at the top of the probe can be
+    marked as outside channels. If "bottom", only the channels at the bottom of the probe can be
+    marked as outside channels. If "both", both the channels at the top and bottom of the probe can be
+    marked as outside channels
+n_neighbors : int, default: 11
+    For coeherence+psd - number of channel neighbors to compute median filter (needs to be odd)
+nyquist_threshold : float, default: 0.8
+    For coherence+psd - frequency with respect to Nyquist (Fn=1) above which the mean of the PSD is calculated and compared
+    with psd_hf_threshold
+direction : "x" | "y" | "z", default: "y"
+    For coherence+psd - the depth dimension
+highpass_filter_cutoff : float, default: 300
+    If the recording is not filtered, the cutoff frequency of the highpass filter
+chunk_duration_s : float, default: 0.5
+    Duration of each chunk
+num_random_chunks : int, default: 100
+    Number of random chunks
+    Having many chunks is important for reproducibility.
+welch_window_ms : float, default: 10
+    Window size for the scipy.signal.welch that will be converted to nperseg
+neighborhood_r2_threshold : float, default: 0.95
+    R^2 threshold for the neighborhood_r2 method.
+neighborhood_r2_radius_um : float, default: 30
+    Spatial radius below which two channels are considered neighbors in the neighborhood_r2 method.
+seed : int or None, default: None
+    The random seed to extract chunks
+
+    """
+
+    def __init__(
+        self,
+        recording: BaseRecording,
+        **detect_bad_channels_kwargs,
+    ):
+
+        # get the default parameters from `detect_bad_channels`, and update with any user-specified parameters.
+        sig = signature(detect_bad_channels)
+        updated_detect_bad_channels_kwargs = {k: v.default for k, v in sig.parameters.items() if k != "recording"}
+        updated_detect_bad_channels_kwargs.update(detect_bad_channels_kwargs)
+
+        bad_channel_ids, channel_labels = detect_bad_channels(recording=recording, **detect_bad_channels_kwargs)
```

Comment:
```suggestion
       bad_channel_ids = detect_bad_channels_kwargs.get("bad_channel_ids")
        if bad_channel_ids is None:
            bad_channel_ids, channel_labels = detect_bad_channels(recording=recording, **detect_bad_channels_kwargs)
```
We don't have to rerun bad channel detecion if bad channels have already been computed. Otherwise, if we parallelize this then each process/thread will rerun detection.

We should also keep the `ProcessingPipeline` in mind. In case we have a processing pipeline, we should differentiate somehow the lazy/non-lazy computation steps.
An option could be to extend the `self._kwargs` mechanism, and add a `self._precomputed_kwargs` (in this case this would have the `bad_channel_ids`).
When loading an object, we could then specify whether we want to use or ignore the precomputed kwargs. When applying the pipeline to another object, you would ignore the precomputed, so that channel detection will be repreformed on the new recording.

Does it make sense?

Reply:
Oh you're very right, great point. For the `ProcessingPipeline` I was thinking that the main distinction is if the parameter is recording-specific or not. If the kwarg (e.g. `bad_channel_ids` or `whitening_matrix` - these might be the only examples??) depends on which recording it's applied to, it shouldn't be used by the pipeline. I reckon this is the exact same set of kwargs that could be precomputed, since you'll always need a recording to do a precomputation?

Reply:
`_kwargs` are handled by the subclasses, so would we need to add `_precomputed_kwargs` to all the preprocessing subclasses?

Reply:
This could be done in the base object. It will need to be there because we will have to handle it upon loading and give the option to add them to the main _kwargs (default), or skip them. @samuelgarcia what do you think?

Reply:
Oh no, we could use `hasattr` to check if a preprocessing class has possible `_precomputed_kwargs`, nice.

Reply:
yes, that is actually probably the best option. But: if we save to json, we need to somehow merge them with the _kwargs, otherwise they get lost!

Reply:
In this case, what happens if someone runs the function, then changes the kwargs (e.g. maybe set them stricter to remove more channels)? would nothing happen on the second call?

Reply:
sorry that wasn't clear I meant the kwargs on the public function like:

```
chan_removed_rec = remove_bad_channels(recording, noisy_channel_threshold=5)

# do some plotting or something to check the remove channels, and decide want to remove a few more

chan_removed_rec = remove_bad_channels(chan_removed_rec, noisy_channel_threshold=1)

```
Ideally people would pass in the original recording but I don't think the above scenario would be so unlikely. Can I warning be thrown?

Reply:
Yeah, having `bad_channel_ids` as a separate argument sounds ok to me.

I think your example wouldn't cause a problem because they are two independent `RemoveBadChannelRecording` classes each with their own kwargs.  Will check...

The difficult bit is that there are two things we want. First, if you load a recording from the `recording.json` file, it should quickly be able to reconstruct the preprocessed recording. For this to work and be fast, we need to save the `bad_channel_ids` in the kwargs and apply them when we load the recording, rather than re-computing.

On the other hand, thinking about the PreprocessingPipeline, we want people to be able to share their pipelines and apply a pipeline from another lab to their own data. To do this, we need to be able to share which kwargs were used to detect the bad channels, but *not* the bad channels themselves; since different recordings have different bad channels.

So we need to mark out that as a special kwarg: apply it when you load the recording, but don't add it to a PreprocessingPipeline. Does that make sense?

Reply:
To summarise the problem (and clarify in my head...)

The difficult bit is that there are two things we want. First, if you load a recording from the `recording.json` file, it should quickly be able to reconstruct the preprocessed recording. For this to work and be fast, we need to save the `bad_channel_ids` in the kwargs and apply them when we load the recording, rather than re-computing.

On the other hand, thinking about the PreprocessingPipeline, we want people to be able to share their pipelines and apply a pipeline from another lab to their own data. To do this, we need to be able to share which kwargs were used to detect the bad channels, but *not* the bad channels themselves; since different recordings have different bad channels.

So we need to mark out that as a special kwarg: apply it when you load the recording, but don't add it to a PreprocessingPipeline. Does that make sense?

Reply:
Yes that makes total sense

Reply:
Hello, I think this is sorted now. We've added `_precomputable_kwarg_names = ["bad_channel_ids"]` and as a class property, and we'll use this on `load` or when making a `pipeline` to include/not include the bad channel ids.

Note: when you load a saved file from dict, it makes an e.g. `DetectAndRemoveBadChannelsRecording` but skips the detection stage, since it knows the `bad_channel_ids`. Hence you can load this recording very fast!